src/tablestore.rs

```diff
@@ -270,15 +276,75 @@ impl TableStore {
         handle: &SSTableHandle,
         index: Arc<SsTableIndexOwned>,
         blocks: Range<usize>,
-    ) -> Result<VecDeque<Block>, SlateDBError> {
+        cache_blocks: bool,
+    ) -> Result<VecDeque<Arc<Block>>, SlateDBError> {
         let path = self.path(&handle.id);
         let obj = ReadOnlyObject {
             object_store: self.object_store.clone(),
             path,
         };
-        self.sst_format
-            .read_blocks(&handle.info, index.as_ref(), blocks, &obj)
-            .await
+        let mut blocks_read = VecDeque::with_capacity(blocks.end - blocks.start);
+        let mut uncached_ranges = Vec::new();
+
+        if let Some(cache) = &self.block_cache {
+            let cached_blocks = join_all(
+                blocks
+                    .clone()
+                    .map(|block_num| async move { cache.get(&(handle.id, block_num)).await }),
+            )
+            .await;
+
+            let mut last_uncached_start = None;
+
+            for (index, block_result) in cached_blocks.into_iter().enumerate() {
+                match block_result.and_then(|b| b.into()) {
+                    Some(cached_block) => {
+                        if let Some(start) = last_uncached_start.take() {
+                            uncached_ranges.push((blocks.start + start)..(blocks.start + index));
+                        }
+                        blocks_read.push_back(cached_block);
+                    }
+                    None => {
+                        last_uncached_start.get_or_insert(index);
+                    }
+                }
+            }
+
+            if let Some(start) = last_uncached_start {
+                uncached_ranges.push((blocks.start + start)..blocks.end);
+            }
+        } else {
+            uncached_ranges.push(blocks.clone());
+        }
+
+        // Read uncached blocks
+        let uncached_blocks = join_all(uncached_ranges.iter().map(|range| {
+            let obj_ref = &obj;
+            let index_ref = &index;
+            async move {
+                self.sst_format
+                    .read_blocks(&handle.info, index_ref, range.clone(), obj_ref)
+                    .await
+            }
+        }))
+        .await;
+
+        // Merge uncached blocks with blocks_read
+        for (range, range_blocks) in uncached_ranges.into_iter().zip(uncached_blocks) {
+            for (block_num, block_read) in range.zip(range_blocks?) {
+                let block = Arc::new(block_read);
+                if cache_blocks {
+                    if let Some(cache) = &self.block_cache {
+                        cache
+                            .insert((handle.id, block_num), CachedBlock::Block(block.clone()))
+                            .await;
```

Comment:
Wonder if it's faster to save the futures in a vec and do a join_all after these for loops?